### PR TITLE
Add wait conditions to combat flakiness

### PR DESCRIPTION
**Pull request by adleong**
We intermittently see flaky policy integration test failures like:

```
 failures:
    either

thread 'either' panicked at 'assertion failed: `(left == right)`
  left: `7`,
 right: `0`: blessed uninjected curl must succeed', policy-test/tests/e2e_server_authorization.rs:293:9
```

This test failure is saying that the curl process is returning an exit code of 7 instead of the expected exit code of 0.  This exit code indicates that curl failed to establish a connection.  https://everything.curl.dev/usingcurl/returns

It's unclear why this connection occasionally fails in CI and I have not been able to reproduce this failure locally.

However, by looking at the logic of the integration test, we can see that the integration test creates the `web` Service and the `web` Pod and waits for that pod to become ready before unblocking the curl from executing.  This means that, theoretically, there could be a race condition between the test and the kubernetes endpoints controller.  As soon as the web pod becomes ready, the endpoints controller will update the endpoints resource for the `web` Service and at the same time, our test will unblock the curl command.  If the test wins this race, it is possible that curl will run before the endpoints resource has been updated.

We add an additional wait condition to the test to wait until the endpoints resource has an endpoint before unblocking curl.

Since I could not reproduce the test failure locally, it is impossible to say if this is actually the cause of the flakiness or if this change fixes it.

### CODE REVIEW

**Comment by adleong** [comment]
Ah, the irony of hitting a *different* test flakiness.  For the record:

```
thread 'inbound_accepted_reconcile_parent_delete' panicked at 'assertion failed: `(left == right)`
  left: `"False"`,
 right: `"True"`', policy-test/tests/inbound_http_route_status.rs:278:9
```

I will rerun the integration tests and investigate that flakiness separately.

File: policy-test/tests/e2e_authorization_policy.rs [hunk 1]
```diff
@@ -294,9 +296,19 @@ async fn network() {
...
         );
 
+        // Wait for the endpoints controller to populate the Endpoints resource.
+        let endpoints_ready = |obj: Option<&k8s::Endpoints>| -> bool {
+            if let Some(ep) = obj {
+                return ep.subsets.iter().flatten().count() > 0;
+            }
+            false
+        };
+        await_condition(&client, &ns, "web", endpoints_ready).await;
```
Comment (hawkw) [inline policy-test/tests/e2e_authorization_policy.rs:306]: nit/TIOLI: is there a reason `endpoints_ready` has to be repeated in all these tests, or can it be a helper in `linkerd_policy_test`?

**Review by hawkw — APPROVED** [review]

File: policy-test/src/lib.rs [hunk 2]
```diff
@@ -168,6 +168,14 @@ pub async fn await_route_status(
     .inner
 }
 
+// Wait for the endpoints controller to populate the Endpoints resource.
+pub fn endpoints_ready(obj: Option<&k8s::Endpoints>) -> bool {
+    if let Some(ep) = obj {
+        return ep.subsets.iter().flatten().count() > 0;
+    }
+    false
```
Comment (hawkw) [inline policy-test/src/lib.rs:176]: nit, take it or leave it: could also be represented as 
```suggestion
   obj.into_iter().flat_map(|ep| ep.subsets.iter().flatten()).count() > 0
```
although :woman_shrugging: if that's any clearer,

File: policy-test/src/lib.rs [hunk 3]
```diff
@@ -168,6 +168,14 @@ pub async fn await_route_status(
     .inner
 }
 
+// Wait for the endpoints controller to populate the Endpoints resource.
```
Comment (hawkw) [inline policy-test/src/lib.rs:171]: nit, not a big deal:
```suggestion
/// Returns `true` if the endpoints controller has populated an Endpoints resource.
```

Comment (hawkw) [reply policy-test/src/lib.rs:171]: we could leave the "// wait for..." comment in the places where this is actually used with `await_condition`, if you want, but not a big deal either way.

**Review by hawkw — APPROVED** [review]
looks good to me! i commented on a couple of nits that you're welcome to address or ignore :)

**Review by alpeb — APPROVED** [review]

**Pull request merged**